src/sign/records.rs

```diff
@@ -243,6 +256,239 @@ impl<N, D> SortedRecords<N, D> {
+        <Octets::Builder as OctetsBuilder>::AppendError: Debug,
+        OctetsMut: OctetsBuilder
+            + AsRef<[u8]>
+            + AsMut<[u8]>
+            + EmptyBuilder
+            + FreezeBuilder,
+    {
+        // TODO:
+        //   - Handle name collisions? (see RFC 5155 7.1 Zone Signing)
+        //   - RFC 5155 section 2 Backwards compatibility:
+        //     Reject old algorithms? if not, map 3 to 6 and 5 to 7, or reject
+        //     use of 3 and 5?
+
+        // RFC 5155 7.1 step 5: _"Sort the set of NSEC3 RRs into hash order."
+        // We store the NSEC3s as we create them in a self-sorting vec.
+        let mut nsec3s = SortedRecords::new();
+
+        // The owner name of a zone cut if we currently are at or below one.
+        let mut cut: Option<FamilyName<N>> = None;
+
+        let mut families = self.families();
+
+        // Since the records are ordered, the first family is the apex --
+        // we can skip everything before that.
+        families.skip_before(apex);
+
+        // We also need the apex for the last NSEC.
+        let apex_owner = families.first_owner().clone();
+        let apex_label_count = apex_owner.iter_labels().count();
+
+        for family in families {
+            // If the owner is out of zone, we have moved out of our zone and
+            // are done.
+            if !family.is_in_zone(apex) {
+                break;
+            }
+
+            // If the family is below a zone cut, we must ignore it.
+            if let Some(ref cut) = cut {
+                if family.owner().ends_with(cut.owner()) {
+                    continue;
+                }
+            }
+
+            // A copy of the family name. We’ll need it later.
+            let name = family.family_name().cloned();
+
+            // If this family is the parent side of a zone cut, we keep the
+            // family name for later. This also means below that if
+            // `cut.is_some()` we are at the parent side of a zone.
+            cut = if family.is_zone_cut(apex) {
+                Some(name.clone())
+            } else {
+                None
+            };
+
+            // RFC 5155 7.1 step 2:
+            //   "If Opt-Out is being used, owner names of unsigned
+            //    delegations MAY be excluded."
+            let has_ds = family.records().any(|rec| rec.rtype() == Rtype::DS);
+            if cut.is_some() && !has_ds && opt_out {
+                continue;
+            }
+
+            // RFC 5155 7.1 step 4:
+            //   "If the difference in number of labels between the apex and
+            //    the original owner name is greater than 1, additional NSEC3
+            //    RRs need to be added for every empty non-terminal between
+            //    the apex and the original owner name."
+            let distance_to_root = name.owner().iter_labels().count();
+            let distance_to_apex = distance_to_root - apex_label_count;
+            if distance_to_apex > 1 {
```

Comment:
True, but the if statement matches nicely with the RFC text and makes it clear that if we enter this block it is because of the condition identified by the RFC text.